lib/Sema/TypeCheckDecl.cpp

```diff
@@ -912,18 +913,81 @@ IsFinalRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
   return explicitFinalAttr;
 }
 
-bool IsMoveOnlyRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
-  // TODO: isNoncopyable and isMoveOnly and other checks are all spread out
-  // and need to be merged together.
+bool IsNoncopyableRequest::evaluate(Evaluator &evaluator, TypeDecl *decl) const {
+  assert(isa<NominalTypeDecl>(decl) && "todo: handle non-nominals");
+
+  // The legacy check.
   if (isa<ClassDecl>(decl) || isa<StructDecl>(decl) || isa<EnumDecl>(decl)) {
-      if (decl->getAttrs().hasAttribute<MoveOnlyAttr>()) {
-        if (!decl->getASTContext().supportsMoveOnlyTypes())
-            decl->diagnose(diag::moveOnly_requires_lexical_lifetimes);
+    if (decl->getAttrs().hasAttribute<MoveOnlyAttr>()) {
+      if (!decl->getASTContext().supportsMoveOnlyTypes())
+          decl->diagnose(diag::moveOnly_requires_lexical_lifetimes);
 
-        return true;
+      return true;
+    }
+  }
+
+  auto &ctx = decl->getASTContext();
+  if (!ctx.LangOpts.hasFeature(Feature::NoncopyableGenerics))
+    return false;
+
+  // For protocols, inspect its requirement signature for a Copyable requirement
+  if (auto *proto = dyn_cast<ProtocolDecl>(decl)) {
+    auto reqs = proto->getRequirementSignature().getRequirements();
```

Comment:
This won't work if `Self: Copyable` is implied by some other `Self: P`. You can instead do this

```
protoDecl->getGenericSignature()->isRequiredProtocol(protoDecl->getSelfInterfaceType(), copyableProto)
```

Reply:
Fixed!